### PR TITLE
Augmentation

### PR DESCRIPTION
**Pull request by anguelos**
This the augmentation API as I though of it for me trying to incorporate things korni.augmentation needs

1)Define augmentation by sample or by batch or both
2)Deterministic augmentaion
3)Augmentation light weight object with its own random seed
4) Factory creates augmentations and inherits torch.nn.Module

### CODE REVIEW

File: kornia/augmentation/tormentor_minimal/base_augmentation.py [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+        functions = (types.BuiltinFunctionType, types.BuiltinMethodType, types.FunctionType)
+        attribute_strings = []
+        for name in self.__dict__.keys():
+            attribute = getattr(self, name)
+            if (not isinstance(attribute, functions)) and (not name.startswith("__")):
+                attribute_strings.append(f"{name}:{repr(attribute)}")
+        return self.__class__.__name__ + ":\n\t" + "\n\t".join(attribute_strings)
+
+
+    def forward_sample(self, tensor_image):
```
Comment (shijianjian) [inline kornia/augmentation/tormentor_minimal/base_augmentation.py]: Do we need forward_sample if we already have forward_batch? As a torch augmentation lib, I think we should encourage people to do their augmentation in batch mode.

File: kornia/augmentation/tormentor_minimal/spatial_augmentations.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+    def forward_sample(self, tensor_image):
+        tensor_image = tensor_image.unsqueeze(dim=0)
+        self.rotate_radians = (torch.rand(1) + self.min_angle)*(self.max_angle-self.min_angle)
+        rotate_degrees = (self.rotate_radians * 180) / math.pi
+        print(rotate_degrees)
+        return kornia.geometry.rotate(tensor_image, rotate_degrees)[0, :, :, :]
+
+
+@aug_parameters(min_scale=.5, max_scale=2.0, joint=False)
```
Comment (shijianjian) [inline kornia/augmentation/tormentor_minimal/spatial_augmentations.py]: I personally do not like the joint parameter. I think we can use max_scale=2.0 to represent joint, max_scale=[2.0, 1.0] to represent non-joint.

File: kornia/augmentation/tormentor_minimal/spatial_augmentations.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+        size_ranges = torch.Tensor([self.max_width - self.min_width, self.max_height - self.min_height])
+        patch_width, patch_height = (torch.rand(2) * size_ranges + size_minima).tolist()
+        left, top = (torch.rand(2)).tolist()
+        left = math.round(image_width - patch_width) * left
+        top = math.round(image_height - patch_height) * top
+        result[:, :, left:left+patch_width, top:top+patch_height] = 0
+        return result
+
+@aug_parameters(desired_width=224, desired_height=224)
+class CropPadAsNeeded(SpatialImageAugmentation):
```
Comment (shijianjian) [inline kornia/augmentation/tormentor_minimal/spatial_augmentations.py]: To handle the padding, I was using:
```python
        if self.pad_if_needed and input.shape[-2] < self.size[0]:
            padding = [0, 0, (self.size[0] - input.shape[-2]), self.size[0] - input.shape[-2]]
            input = F.pad(input, padding, value=self.fill, mode=self.padding_mode)

        if self.pad_if_needed and input.shape[-1] < self.size[1]:
            padding = [self.size[1] - input.shape[-1], self.size[1] - input.shape[-1], 0, 0]
            input = F.pad(input, padding, value=self.fill, mode=self.padding_mode)
```
I think it looks more cleaner.

File: kornia/augmentation/tormentor_minimal/base_augmentation.py (hunk 1 above)
Comment (anguelos) [reply kornia/augmentation/tormentor_minimal/base_augmentation.py]: Encourage yes but why force?

File: kornia/augmentation/tormentor_minimal/spatial_augmentations.py (hunk 2 above)
Comment (anguelos) [reply kornia/augmentation/tormentor_minimal/spatial_augmentations.py]: That is definetly more elegant. A proper typeunion for valid random ranges should be broadly used IMO.

File: kornia/augmentation/tormentor_minimal/spatial_augmentations.py (hunk 3 above)
Comment (anguelos) [reply kornia/augmentation/tormentor_minimal/spatial_augmentations.py]: Actually pad_if_needed was my only contribution to torchvision :D.
Whether pad_if_needed is a switch or a different augmentation to me seems secondary. If torchvision compatibillity is desired, I guess pad_if_needed is better.
Should everything be padded the same way? If augmenting, ideally padding borders should be random adding up to what is needed.

File: kornia/augmentation/tormentor_minimal/base_augmentation.py [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+            self.id = id
+        if seed is None:
+            self.seed = self.id + SpatialImageAugmentation._global_seed
+        else:
+            self.seed = seed
+
+    def __call__(self, tensor_image):
+        device = tensor_image.device
+        with torch.random.fork_rng(devices=(device,)):
+            torch.manual_seed(self.seed)
```
[the remark this answers is not in the document]
Comment (anguelos) [reply kornia/augmentation/tormentor_minimal/base_augmentation.py]: For me forking the RNG provides absolute determinism regardless of what the children classes did as long as they stay in a single thread/process don't use foreign random generators eg. numpy.
People writing an augmentation use anything they want torch.rand etc, it also plays well with torch.distributions.
On my machine it takes 780 ns on torch.random.fork_rng(devices=["cuda","cpu"])
This would be an added cost per __call__. In the use case I have in mind this would be acceptable, by batch augmentation makes it even cheaper.

File: kornia/augmentation/tormentor_minimal/base_augmentation.py (hunk 1 above)
Comment (anguelos) [reply kornia/augmentation/tormentor_minimal/base_augmentation.py]: The by-sample approach has some advantages and some disadvantages.
It allows for images of different sizes to be augmented on their own; A Final CropPadAsNeeded is a good way to merge images of heterogeneous sizes into a single batch after they were augmented. 
The later down a pipeline images have to conform to a uniform size, the more flexible the pipeline is.

File: kornia/augmentation/tormentor_minimal/base_augmentation.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (edgarriba) [reply kornia/augmentation/tormentor_minimal/base_augmentation.py]: basically what I understand here is to decided using `torch.random.fork` vs taking global `torch.Generator`, right ? @shijianjian how's your stuff in terms of memory footprint, @anguelos approach seems quite light-weighted just storing the random seed with just an integer.

File: kornia/augmentation/tormentor_minimal/base_augmentation.py (hunk 1 above)
Comment (edgarriba) [reply kornia/augmentation/tormentor_minimal/base_augmentation.py]: I agree with @shijianjian probably for our use cases which all the operators in `kornia` are meant for batch this won't make sense at all. If you want to run per sample you could unsqueeze one dim and make B==1 and then go back again.

File: kornia/augmentation/tormentor_minimal/spatial_augmentations.py (hunk 2 above)
Comment (edgarriba) [reply kornia/augmentation/tormentor_minimal/spatial_augmentations.py]: tuple might be more suitable in this context

File: kornia/augmentation/tormentor_minimal/base_augmentation.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (anguelos) [reply kornia/augmentation/tormentor_minimal/base_augmentation.py]: The main reason for forking the random seed is thread-safety/determinism. In a single thread a cascade of torch.rand, will always be the same for the same seed but if you have multi-thread parallelism you lose determinism. Also careful if you fork, the two forks might end up doing exactly the same random stuff.

File: kornia/augmentation/tormentor_minimal/spatial_augmentations.py (hunk 2 above)
Comment (anguelos) [reply kornia/augmentation/tormentor_minimal/spatial_augmentations.py]: I my new draft (), I think a light wrapper around torch.distributions would be even more appropriate. Simething like Uniform(.5,2.0).
